### ISSUE TITLE
PERF: timezoned series created 6x faster than non-timezoned series

### ISSUE THREAD

**Issue by soerenwolfers** [pandas-dev/pandas#56860]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this issue exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this issue exists on the main branch of pandas.

### Reproducible Example

In a jupyter notebook
```
import pandas as pd
import numpy as np
n = 1_000_000
%time a = pd.Series(np.repeat(pd.Timestamp('2019-01-01 00:00:00', tz=None), n))
%time b = pd.Series(np.repeat(pd.Timestamp('2019-01-01 00:00:00', tz='UTC'), n))
```
```
PU times: user 1.88 s, sys: 11.9 ms, total: 1.89 s
Wall time: 1.88 s
CPU times: user 315 ms, sys: 3.99 ms, total: 319 ms
Wall time: 319 ms
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit              : a671b5a8bf5dd13fb19f0e88edc679bc9e15c673
python              : 3.10.6.final.0
python-bits         : 64
OS                  : Linux
OS-release          : 5.15.0-76-generic
Version             : #83-Ubuntu SMP Thu Jun 15 19:16:32 UTC 2023
machine             : x86_64
processor           : x86_64
byteorder           : little
LC_ALL              : en_GB.utf8
LANG                : C.UTF-8
LOCALE              : en_GB.UTF-8

pandas              : 2.1.4
numpy               : 1.24.4
pytz                : 2023.3
dateutil            : 2.8.2
setuptools          : 68.0.0
pip                 : 23.2.1
Cython              : None
pytest              : None
hypothesis          : None
sphinx              : None
blosc               : None
feather             : None
xlsxwriter          : None
lxml.etree          : None
html5lib            : None
pymysql             : None
psycopg2            : None
jinja2              : 3.1.2
IPython             : 8.14.0
pandas_datareader   : None
bs4                 : 4.12.2
bottleneck          : None
dataframe-api-compat: None
fastparquet         : None
fsspec              : None
gcsfs               : None
matplotlib          : 3.7.2
numba               : 0.57.1
numexpr             : None
odfpy               : None
openpyxl            : None
pandas_gbq          : None
pyarrow             : 14.0.2
pyreadstat          : None
pyxlsb              : None
s3fs                : None
scipy               : 1.11.1
sqlalchemy          : None
tables              : None
tabulate            : None
xarray              : None
xlrd                : None
zstandard           : 0.21.0
tzdata              : 2023.3
qtpy                : 2.3.1
pyqt5               : None

</details>

**Comment by soerenwolfers** [comment]
disappears after upgrading to `pyarrow==14.0.2`

**Comment by rhshadrach** [comment]
Thanks for the report, confirmed on main. Almost all the time is being spent in Series construction. Further investigations are welcome.

**Comment by jrmylow** [comment]
I've started poking into this, posting a summary note before I pick this up later:
* Using `cProfile` the big difference comes up in the call to `objects_to_datetime64` here https://github.com/pandas-dev/pandas/blob/e3796929283c91f0f96fed3d37b2e85bf4007ef9/pandas/core/arrays/datetimes.py#L2368
* This calls out to Cython code next so it doesn't get picked up by cProfile https://github.com/pandas-dev/pandas/blob/e3796929283c91f0f96fed3d37b2e85bf4007ef9/pandas/_libs/tslib.pyx#L414
* The key difference in the two cases appears to happen in the next call, which is `conversion.pyx`. 
https://github.com/pandas-dev/pandas/blob/e3796929283c91f0f96fed3d37b2e85bf4007ef9/pandas/_libs/tslibs/conversion.pyx#L773
* Specifically, because case b's values contain the `tzinfo` attribute, it calls out to `convert_datetime_to_tsobject` while case a calls out to `(<_Timestamp>val)._as_creso(creso, round_ok=True)._value`

At this stage I probably still need to find a way to confirm if there is a significant performance difference between these branches, and whether there is a reasonable fix.

**Comment by jrmylow** [comment]
If I had to guess - it would seem that the `_TSObject` that is created in case b is lighter/faster than the `_Timestamp` created in case a?

**Comment by mroeschke** [comment]
The difference is in the resolution conversion to `ns`

```
In [13]: %time a = pd.Series(np.repeat(pd.Timestamp('2019-01-01 00:00:00', tz=None), n))
CPU times: user 809 ms, sys: 6.04 ms, total: 815 ms
Wall time: 813 ms

In [14]: %time a = pd.Series(np.repeat(pd.Timestamp('2019-01-01 00:00:00', tz=None).as_unit("ns"), n))
CPU times: user 344 ms, sys: 3.31 ms, total: 347 ms
Wall time: 347 ms
```

AFAICT, the non-tz case actually does a resolution conversion from "s" to "ns" while the tz case just creates a new object "replacing" the "s" resolution with "ns".

Technically, both should return "s" resolution here.

**Comment by jbrockmendel** [comment]
TL;DR There turns out to be a missing `break` statement inside `lib.maybe_convert_objects`.  Posting the rest of the tracking-this-down process for pedagogical purposes:

Drilling down on where the time is spent (`%prun` isn't giving very useful info for me, i suspect bc most of the time is spent in cython)

```
n = 10**6
arr1 = np.repeat(pd.Timestamp('2019-01-01 00:00:00', tz=None), n)
arr2 = np.repeat(pd.Timestamp('2019-01-01 00:00:00', tz='UTC'), n)

In [75]: %timeit a = pd.Series(arr1)
359 ms ± 1.47 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [76]: %timeit b = pd.Series(arr2)
198 ms ± 2.26 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

So its only 1.8x for me, which is a lot better than 6x.  Drilling down further:

```
In [77]: %timeit a = pd.core.construction.sanitize_array(arr1, None)
367 ms ± 17.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [78]: %timeit b = pd.core.construction.sanitize_array(arr2, None)
196 ms ± 1.62 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Which confirms that effectively all the time is spent inside sanitize_array.  More:

```
In [80]: %timeit a = pd._libs.lib.maybe_convert_objects(arr1, convert_non_numeric=True)
356 ms ± 6.03 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [81]: %timeit b = pd._libs.lib.maybe_convert_objects(arr2, convert_non_numeric=True)
198 ms ± 797 μs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Again effectively all the time.  Next, inside maybe_convert_objects these go down different paths, paraphrasing:

```
    if seen.datetimetz_:  # <- tzaware case
        if is_datetime_with_singletz_array(objects):  
    elif seen.datetime_:  # <- tznaive case
        if is_datetime_or_datetime64_array(objects):
```

So we guess that either the DatetimeIndex call is the difference, or the is_datetime_yadda_yadda.  To check this we need to change the `cdef is_datetime_or_datetime64_array` to cpdef, then:

```
In [6]: %timeit pd._libs.lib.is_datetime_with_singletz_array(arr2)
94.4 ms ± 1.53 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%timeit pd._libs.lib.is_datetime_or_datetime64_array(arr1)
93.4 ms ± 839 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Then the DatetimeIndex constructor:

```
In [82]: %timeit a = pd.DatetimeIndex(arr1)
52.7 ms ± 1.16 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [83]: %timeit b = pd.DatetimeIndex(arr2)
106 ms ± 453 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

OK, we're on to something!  the is_datetime_foo calls were the same speed and the DatetimeIndex calls were actually _faster_ for the tznaive case.  So something must be amiss inside lib.maybe_convert_objects _before_ we get to the is_datetime_foo calls.  Looking up at the code that sets `seen.datetimetz` and `seen.datetime_`:

```
        elif PyDateTime_Check(val) or cnp.is_datetime64_object(val):

            # if we have a tz's attached then return the objects
            if convert_non_numeric:
                if getattr(val, "tzinfo", None) is not None:
                    seen.datetimetz_ = True
                    break
                else:
                    seen.datetime_ = True
                    try:
                        convert_to_tsobject(val, None, None, 0, 0)
                    except OutOfBoundsDatetime:
                        # e.g. test_out_of_s_bounds_datetime64
                        seen.object_ = True
                        break
```

The alert reader will note a missing `break` in the tznaive not-of-bounds case.  Adding that `break`:

```
In [4]: %timeit a = pd.Series(arr1)
152 ms ± 2.12 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [5]: %timeit b = pd.Series(arr2)
211 ms ± 9.89 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Ta-da!
